osgar/explore.py

```diff
@@ -0,0 +1,146 @@
+    return None
+
+
+class FollowWall(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+
+    def send_speed_cmd(self, speed, angular_speed):
+        return self.publish('desired_speed', [round(speed*1000), round(math.degrees(angular_speed)*100)])
+
+    def update(self):  # hack, this method should be called run instead!       
+        channel = super().update()  # define self.time
+
+        moving = False
+        desired_speed = 0.0
+        start_time = self.time
+        while self.time - start_time < timedelta(minutes=1):
+            channel = channel = super().update()
+            if channel == 'scan':
+                assert len(self.scan) == 811, len(self.scan)  # 541
+                distL, distR = min_dist_arr(self.scan[270:-270])  # was 200
+                distL = 20.0 if distL is None else distL
+                distR = 20.0 if distR is None else distR
+                dist = min(distL, distR)
+                desired_angular_speed = follow_wall_angle(self.scan, radius=WALL_DISTANCE)
+                print(desired_angular_speed)
+                if moving:
+                    if dist is None or dist < SAFE_DISTANCE_STOP:
+                        print("!!! STOP !!!",  dist, (distL, distR))
+                        desired_speed = 0.0
+                        desired_angular_speed = 0.0
+                        moving = False
+                else:  # not moving
+                    if dist > SAFE_DISTANCE_GO:
+                        print("GO",  dist)
+                        desired_speed = DESIRED_SPEED
+                        moving = True
+                self.send_speed_cmd(desired_speed, desired_angular_speed)
+
+            elif channel == 'emergency_stop':
+                self.send_speed_cmd(0.0, 0.0)
+                self.request_stop()  # it should be "delayed"
+
+
+if __name__ == "__main__":
+    from osgar.logger import LogWriter, LogReader
+    import argparse
+    import sys
+
+    from osgar.lib.config import load as config_load
+    from osgar.record import Recorder
+
+    parser = argparse.ArgumentParser(description='Follow Wall')
+    subparsers = parser.add_subparsers(help='sub-command help', dest='command')
+    subparsers.required = True
+    parser_run = subparsers.add_parser('run', help='run on real HW')
+    parser_run.add_argument('config', nargs='+', help='configuration file')
+    parser_run.add_argument('--note', help='add description')
+
+    parser_replay = subparsers.add_parser('replay', help='replay from logfile')
+    parser_replay.add_argument('logfile', help='recorded log file')
+    parser_replay.add_argument('--force', '-F', dest='force', action='store_true', help='force replay even for failing output asserts')
+    parser_replay.add_argument('--config', nargs='+', help='force alternative configuration file')
+    args = parser.parse_args()
+
+    if args.command == 'replay':
+        from osgar.replay import replay
```

Comment:
Either move this up where you import `Recorder` or move the `Recorder` import down to the 'run' section for consistency. Or is there any hidden reason to have it this way?

Reply:
fixed with launcher